## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [4]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
0                1  
1                0  
2                6  
3                3  
4                1  
...            ...  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5  

[54639 rows x 18 columns]

In [5]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [6]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [7]:
# Use this space to do any additional research
# and familiarize yourself with the data.
#DROP THE NULLS
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [8]:
# What three item categories had the most entries?
category_ordered_count = [df['category'].value_counts().index.tolist(), df['category'].value_counts().tolist()]
# category_ordered_count
print(f"The three items categories that had the most entries are \n1.{category_ordered_count[0][0]} with {category_ordered_count[1][0]} entries, \n2.{category_ordered_count[0][1]} with {category_ordered_count[1][1]}, and \n3.{category_ordered_count[0][2]} with {category_ordered_count[1][2]}.")

The three items categories that had the most entries are 
1.consumables with 23538 entries, 
2.furniture with 11915, and 
3.software with 8400.


In [9]:
# For the category with the most entries, which subcategory had the most entries?
mode_category_filtered_df = df.loc[df['category'] == str(category_ordered_count[0][0])]
# mode_category_filtered_df
print(f"The subcategory of the category '{category_ordered_count[0][0]}' with the most entries is {list(mode_category_filtered_df['subcategory'].mode())[0]}.")

The subcategory of the category 'consumables' with the most entries is bathroom supplies.


In [10]:
# Which five clients had the most entries in the data?


client_ordered_count = [df['client_id'].value_counts().index.tolist(), df['client_id'].value_counts().tolist()]


print("The five clients with the most entries are: ")
for i in range (0,5):
    print(f"{i+1}. {client_ordered_count[0][i]} - {client_ordered_count[1][i]}")






# client_ordered_count = df['client_id'].value_counts().rename_axis('client_id').reset_index(name='counts')
# print(client_ordered_count)
# print(client_ordered_count_df)
# client_ordered_count_df = pd.DataFrame(client_ordered_count, columns = ['client_id', 'orders'])

The five clients with the most entries are: 
1. 33615 - 220
2. 66037 - 211
3. 46820 - 209
4. 38378 - 207
5. 24741 - 207


In [11]:
# Store the client ids of those top 5 clients in a list.
# top_five_clients = client_ordered_count['client_id'].head().tolist()
# # print(top_five_clients)
# top_five_clients_count = client_ordered_count['counts'].head().tolist()
# print(top_five_clients_count)


##PLEASE READ##
##PLEASE READ##
##PLEASE READ##
#                       DONE IN PREVIOUS STEP SINCE THIS BLOCK CONTAINS INEFFICIENT CODE
##PLEASE READ##
##PLEASE READ##
##PLEASE READ##
##PLEASE READ##

In [12]:
# How many total units (the qty column) did the client with the most entries order order?

# df.loc[df['client_id'] == client_ordered_count[0][0], 'qty'].sum()
print(f"Client {client_ordered_count[0][0]} ordered a total of {df.loc[df['client_id'] == client_ordered_count[0][0], 'qty'].sum()} units.")

Client 33615 ordered a total of 64313 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [14]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
def subtotal(row):
    return row['unit_price'] * row['qty']
df['Subtotal'] = df.apply(subtotal, axis=1)

In [15]:
df[['unit_price','Subtotal']].head()

unit_price   Subtotal
0     1096.80  115164.00
1       24.95     523.95
2       13.52     527.28
3       36.42    1056.18
4      195.10    3902.00

In [16]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def shipping(row):
    if row['unit_weight'] * row['qty'] > 50:
        return 7 * row['unit_weight'] * row['qty']
    return 10 * row['unit_weight'] * row['qty']

df['shipping'] = df.apply(shipping, axis=1)

In [17]:
df[['unit_price','Subtotal','shipping']].head()

unit_price   Subtotal  shipping
0     1096.80  115164.00   5512.50
1       24.95     523.95    312.90
2       13.52     527.28    458.64
3       36.42    1056.18    356.70
4      195.10    3902.00   6500.20

In [18]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
def total_price(row):
    total_price = row['Subtotal'] + row['shipping']
    total_price += total_price * (9.25/100)
    return total_price

df['total_price'] = df.apply(total_price, axis=1)

In [19]:
df[['unit_price','Subtotal','shipping','total_price']].head()

unit_price   Subtotal  shipping    total_price
0     1096.80  115164.00   5512.50  131839.076250
1       24.95     523.95    312.90     914.258625
2       13.52     527.28    458.64    1077.117600
3       36.42    1056.18    356.70    1543.571400
4      195.10    3902.00   6500.20   11364.403500

In [20]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

def s_subtotal(row):
    return (row['unit_cost'] * row['qty']) + row['shipping']

df['unit_total'] = df.apply(s_subtotal, axis=1)

In [21]:
df[['unit_price','Subtotal','shipping','total_price','unit_total']].head()

unit_price   Subtotal  shipping    total_price  unit_total
0     1096.80  115164.00   5512.50  131839.076250    85597.05
1       24.95     523.95    312.90     914.258625      629.79
2       13.52     527.28    458.64    1077.117600      765.18
3       36.42    1056.18    356.70    1543.571400     1077.35
4      195.10    3902.00   6500.20   11364.403500     8663.60

In [22]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['unit_total']


In [23]:
df[['unit_price','Subtotal','shipping','total_price','unit_total','profit']].head()

unit_price   Subtotal  shipping    total_price  unit_total        profit
0     1096.80  115164.00   5512.50  131839.076250    85597.05  46242.026250
1       24.95     523.95    312.90     914.258625      629.79    284.468625
2       13.52     527.28    458.64    1077.117600      765.18    311.937600
3       36.42    1056.18    356.70    1543.571400     1077.35    466.221400
4      195.10    3902.00   6500.20   11364.403500     8663.60   2700.803500

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [25]:
# Check your work using the totals above
print(f"Email receipts confirm that the total prices of the 3 orders are: ")

print(f"Order ID 2742071 had a total price of ${df[df['order_id'] == 2742071]['total_price'].sum():,.2F} ")
print(f"Order ID 2173913 had a total price of ${df[df['order_id'] == 2173913]['total_price'].sum():,.2F} ")
print(f"Order ID 6128929 had a total price of ${df[df['order_id'] == 6128929]['total_price'].sum():,.2F} ")

Email receipts confirm that the total prices of the 3 orders are: 
Order ID 2742071 had a total price of $152,811.90 
Order ID 2173913 had a total price of $162,388.72 
Order ID 6128929 had a total price of $923,441.24 


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [27]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

## QUESTION AMBIGUOUS, USING CLIENT ID'S FROM PART 1

for i in range(0,5):
    print(f"{i}. Client {client_ordered_count[0][i]} spent a total of ${df.loc[df['client_id'] == client_ordered_count[0][i], 'total_price'].sum():,.2f}\n")
    

0. Client 33615 spent a total of $8,377,308.57

1. Client 66037 spent a total of $10,259,514.80

2. Client 46820 spent a total of $9,743,794.32

3. Client 38378 spent a total of $12,906,550.88

4. Client 24741 spent a total of $82,268,891.98



In [28]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.

top_client_list=[]

for i in range(0,5):
    client_dict = {
        'client_id' : client_ordered_count[0][i],
        'total units purchased' : df[df['client_id'] == client_ordered_count[0][i]]['qty'].sum(),
        'total shipping price' : df[df['client_id'] == client_ordered_count[0][i]]['shipping'].sum(),
        'total revenue' : df[df['client_id'] == client_ordered_count[0][i]]['Subtotal'].sum(),
        'total profit' : df[df['client_id'] == client_ordered_count[0][i]]['profit'].sum()
    }
    top_client_list.append(client_dict)

In [29]:
top_client_list_df = pd.DataFrame(top_client_list)

In [30]:
top_client_list_df

client_id  total units purchased  total shipping price  total revenue  \
0      33615                  64313            1828984.89     5839032.11   
1      66037                  43018            1395151.85     7995708.38   
2      46820                  75768            1601448.84     7317356.03   
3      38378                  73667            3429455.40     8384321.15   
4      24741                 239862            5126448.37    70176885.25   

   total profit  
0  2.201995e+06  
1  3.255032e+06  
2  2.736603e+06  
3  3.271830e+06  
4  3.657999e+07

In [31]:
# Format the data and rename the columns to names suitable for presentation.
top_client_list_df.rename(columns = {"client_id":"Client ID",
                                     "total units purchased":"Units",
                                     "total shipping price":"Shipping",
                                     "total revenue":"Revenue",
                                     "total profit":"Profit"}, inplace=True)


In [32]:
top_client_list_df

Client ID   Units    Shipping      Revenue        Profit
0      33615   64313  1828984.89   5839032.11  2.201995e+06
1      66037   43018  1395151.85   7995708.38  3.255032e+06
2      46820   75768  1601448.84   7317356.03  2.736603e+06
3      38378   73667  3429455.40   8384321.15  3.271830e+06
4      24741  239862  5126448.37  70176885.25  3.657999e+07

In [33]:
# Define the money columns. 
money_columns = ['Shipping','Revenue','Profit']


In [34]:
# Define a function that converts a dollar amount to millions.
def currency_format_millions(value):
    return round(value/1000000,1)


In [35]:
# Apply the currency_format_millions function to only the money columns. 
for col in money_columns:
    top_client_list_df[col] = top_client_list_df[col].apply(currency_format_millions)


In [36]:
# Rename the columns to reflect the change in the money format. 

top_client_list_df.rename(columns = {"Shipping":"($M)Shipping","Revenue":"($M)Revenue","Profit":"($M)Profit"}, inplace=True)


In [37]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
top_client_list_df.sort_values(by ="($M)Profit", ascending=False, inplace=True)

In [38]:
top_client_list_df.set_index("Client ID")

Units  ($M)Shipping  ($M)Revenue  ($M)Profit
Client ID                                               
24741      239862           5.1         70.2        36.6
66037       43018           1.4          8.0         3.3
38378       73667           3.4          8.4         3.3
46820       75768           1.6          7.3         2.7
33615       64313           1.8          5.8         2.2

### Summary

- Client 24741 had the highest profit of \\$36.6M, however they only had that high of a profit because they brought in \\$70.2M in revenue. Their revenue was supported by the fact that they sold 239862 units which is significatnly higher than any other client.

- Although client 66037 spent the least in shipping, they had the second highest overall profit as their revenue margin was high compared to their unit weight which was low enough to keep their profits high.

- Client 38378 had really high shipping cost compared to the rest of the other clients. This made their overall profit scale lower as they ended up paying more for shipping.

- Client 46820 and 33615 had standard profit scaling margins for their shipping and revenue, but for comparison 46280 sold more units at lower weight getting their shipping cost lower and overall bringing higher profits than 33615.